src/coap_net.c

```diff
@@ -4435,6 +4435,42 @@ coap_dispatch(coap_context_t *context, coap_session_t *session,
       coap_log_debug("Decrypted PDU\n");
       coap_show_pdu(COAP_LOG_DEBUG, pdu);
     }
+  } else if (COAP_PDU_IS_RESPONSE(pdu) &&
+             session->oscore_encryption &&
+             pdu->type != COAP_MESSAGE_RST) {
+    if (COAP_RESPONSE_CODE_BAD_OPTION != pdu->code &&
+        COAP_RESPONSE_CODE_UNAUTHORIZED != pdu->code &&
+        COAP_RESPONSE_CODE_BAD_REQUEST != pdu->code) {
```

Comment:
As per https://github.com/obgm/libcoap/pull/571#issuecomment-727811702 and https://github.com/obgm/libcoap/pull/571#issuecomment-728191913 in #571 this should read
````
   if ( COAP_RESPONSE_CODE(402) != pdu->code &&
        COAP_RESPONSE_CODE(401) != pdu->code &&
        COAP_RESPONSE_CODE(400) != pdu->code) {
````
or preferably as the way the rest of the coding style does it
````
    if (pdu->code != COAP_RESPONSE_CODE(402) &&
        pdu->code != COAP_RESPONSE_CODE(401) &&
        pdu->code != COAP_RESPONSE_CODE(400) {
````
or even better
````
    if (COAP_RESPONSE_CLASS(response->code) == 2) {
````
as 4.xx and 5.xx can be unprotected.  See [RFC 8613 B.2.1. Implementation Considerations](https://datatracker.ietf.org/doc/html/rfc8613#appendix-B.2.1).  This should be OK for the first CSM [RFC 8163 4.3. Signalling Messages](https://datatracker.ietf.org/doc/html/rfc8613#section-4.3), but needs testing.

Reply:
Good to see a greater picture. I agree, that response classes 5 and 7 should also be accepted, but I wander if all 4 class response codes should be accepted, or only the mentioned ones. So I am thinking about following change:
 ```suggestion
    if (pdu->code != COAP_RESPONSE_CODE(402) &&
        pdu->code != COAP_RESPONSE_CODE(401) &&
        pdu->code != COAP_RESPONSE_CODE(400) &&
        COAP_RESPONSE_CLASS(pdu->code) < 5) {
```

The one you suggested looks better:
```suggestion
    if (COAP_RESPONSE_CLASS(response->code) == 2) {
```
but I am not sure if there are some security considerations that should be taken into account. I have to give it a thought.

Reply:
4.08 [RFC9177 5. [The Use of the 4.08 (Request Entity Incomplete) Response Code](https://datatracker.ietf.org/doc/html/rfc9177#name-the-use-of-the-408-request-)](https://datatracker.ietf.org/doc/html/rfc9177#section-5) is another one that should not be checked for.  It is likely that there could be others and new 4.xx may get introduced in new RFCs.

So, again, I think the only safe answer here is
````
     if (COAP_RESPONSE_CLASS(response->code) == 2) {
````
Which catches a 'good' response being used maliciously.

Reply:
I see, let's stay then with the check for response class 2.